File: ios/Classes/ProvisioningBridge.swift
// ProvisioningBridge.swift
//
// Isolates every interaction with Espressif's ESPProvision SDK from the
// Flutter plugin entry point. The plugin class translates method-channel
// invocations into Bridge calls; the Bridge owns the SDK-side state machine,
// emits lifecycle events through the supplied EventEmitter, and translates
// the SDK's error vocabulary into the typed FlutterError codes that the Dart
// side's `mapPlatformException` knows how to dispatch on.
//
// Threading: every public method on Bridge is expected to be called from the
// Flutter UI thread (the method channel is bound there). All Result and
// Event emissions are themselves dispatched to the main queue, since
// ESPProvision callbacks can land on arbitrary queues.

import CoreBluetooth
import ESPProvision
import Flutter
import Foundation

/// Receives lifecycle events from the bridge. Implemented by the plugin
/// class so the bridge stays decoupled from the FlutterEventChannel itself
/// (and so a fake can be substituted in unit tests).
protocol EspEventEmitter: AnyObject {
    func emit(_ event: [String: Any?])
}

/// The username ESP-IDF stamps into security2-enabled firmware by default.
/// Production fleets may override this via firmware; we expose no
/// per-device override because every device we ship uses the default. If
/// that changes, plumb a `username` parameter through to `connect`.
private let kDefaultSecurity2Username = "wifiprov"

/// Wraps ESPProvision's stateful, callback-driven API behind a single
/// per-plugin actor. Holds the discovered device cache, the currently
/// connected device, and the in-flight scan state. Not thread-safe on its
/// own — relies on being called only from the main queue.
final class ProvisioningBridge {

    // MARK: - State

    private weak var eventEmitter: EspEventEmitter?

    /// Last successful scan's results, keyed by `ESPDevice.name`. ESPDevice
    /// instances must be retained between `scanBleDevices` and `connect`,
    /// otherwise the SDK can't re-use the BLE transport that found them.
    private var discoveredDevices: [String: ESPDevice] = [:]

    /// The currently-connected device. Held strong while a session is open
    /// so that ESPProvision's internal queues retain access to it; released
    /// on `disconnect` to avoid retain cycles via the SRP / security layers.
    private var connectedDevice: ESPDevice?

    /// Security level used for the active connection — needed when mapping
    /// `ESPSessionError.sessionInitError` back to a typed Dart exception
    /// (the same SDK error means "wrong PoP" under sec1/sec2 but a generic
    /// handshake failure under sec0).
    private var activeSecurity: ESPSecurity = .secure2

    /// Strong references to per-connection PoP delegates. The SDK holds the
    /// delegate weakly during `connect`; without this we'd lose the closure
    /// before the SDK calls back.
    private var popDelegateRefs: [ObjectIdentifier: PopDelegate] = [:]

    // Concurrency guards.
    private var scanInFlight = false
    private var connectInFlight = false
    private var scanTimer: Timer?
    private var scanResult: FlutterResult?

    // Bluetooth state probe — created lazily, kept alive for the lifetime
    // of the bridge so the iOS Core Bluetooth runtime sees a stable owner.
    private lazy var btProbe = BluetoothStateProbe()

    init(eventEmitter: EspEventEmitter) {
        self.eventEmitter = eventEmitter
    }

    // MARK: - Method channel entry points

    func scanBleDevices(
        devicePrefix: String,
        timeoutMs: Int,
        result: @escaping FlutterResult
    ) {
        // Cancel any previous in-flight scan.
        if scanInFlight {
            ESPProvisionManager.shared.stopESPDevicesSearch()
            stopScanTimer()
            // The previous caller waits for an empty result rather than an
            // exception — superseding scans is a normal UI pattern.
            scanResult?([])
            scanResult = nil
            scanInFlight = false
            emit(phase: "scanFinished", message: "superseded by new scan")
        }

        btProbe.checkAvailability { [weak self] availability in
            guard let self = self else { return }
            DispatchQueue.main.async {
                switch availability {
                case .available:
                    self.startBleScan(devicePrefix: devicePrefix,
                                      timeoutMs: timeoutMs,
                                      result: result)
                case .poweredOff:
                    result(FlutterError(
                        code: "ble_unavailable",
                        message: "Bluetooth is powered off or unsupported on this device",
                        details: nil))
                case .unauthorized:
                    result(FlutterError(
                        code: "permission_denied",
                        message: "App is not authorised to use Bluetooth",
                        details: ["permission": "bluetooth"]))
                }
            }
        }
    }

    func stopBleScan(result: @escaping FlutterResult) {
        if scanInFlight {
            ESPProvisionManager.shared.stopESPDevicesSearch()
            stopScanTimer()
            scanResult?([])
            scanResult = nil
            scanInFlight = false
            emit(phase: "scanFinished", message: "stopped by caller")
        }
        result(nil)
    }

    func connect(
        deviceMap: [String: Any?],
        proofOfPossession: String,
        security: Int,
        result: @escaping FlutterResult
    ) {
        if connectInFlight || connectedDevice != nil {
            result(FlutterError(
                code: "session_failed",
                message: "A connection is already in flight or established; call disconnect() first",
                details: nil))
            return
        }

        guard let deviceId = deviceMap["id"] as? String, !deviceId.isEmpty else {
            result(FlutterError(
                code: "session_failed",
                message: "connect: device map missing valid 'id'",
                details: nil))
            return
        }

        guard let device = discoveredDevices[deviceId] else {
            result(FlutterError(
                code: "device_not_found",
                message: "Device '\(deviceId)' is no longer in the scan cache. Re-scan and try again.",
                details: ["deviceId": deviceId]))
            return
        }

        let espSecurity = ESPSecurity(rawValue: security)
        device.security = espSecurity
        // `proofOfPossession` is `internal` on ESPDevice — we cannot set it
        // directly. The PopDelegate below supplies it through the public
        // delegate callback that ESPDevice.initialiseSession invokes when
        // its own `proofOfPossession` slot is nil (which it always is for
        // devices freshly discovered by `searchESPDevices`).
        if espSecurity == .secure2 {
            device.username = kDefaultSecurity2Username
        }
        activeSecurity = espSecurity
        connectInFlight = true

        let popDelegate = PopDelegate(pop: proofOfPossession,
                                      username: kDefaultSecurity2Username)
        let popKey = ObjectIdentifier(popDelegate)
        popDelegateRefs[popKey] = popDelegate

        emit(phase: "connecting", deviceId: deviceId)

        device.connect(delegate: popDelegate) { [weak self] status in
            guard let self = self else { return }
            DispatchQueue.main.async {
                self.popDelegateRefs.removeValue(forKey: popKey)
                self.connectInFlight = false
                switch status {
                case .connected:
                    self.connectedDevice = device
                    self.emit(phase: "sessionEstablished", deviceId: deviceId)
                    result(nil)
                case .failedToConnect(let error):
                    self.emit(phase: "disconnected",
                              deviceId: deviceId,
                              message: error.description)
                    result(ErrorMapping.flutterError(
                        forSessionError: error,
                        security: self.activeSecurity,
                        ssidOrName: deviceId))
                case .disconnected:
                    self.emit(phase: "disconnected", deviceId: deviceId)
                    result(FlutterError(
                        code: "session_failed",
                        message: "Connection terminated before session established",
                        details: nil))
                }
            }
        }

        emit(phase: "sessionEstablishing", deviceId: deviceId)
    }

    func scanWifiNetworks(result: @escaping FlutterResult) {
        guard let device = connectedDevice else {
            result(FlutterError(
                code: "session_failed",
                message: "No active session. Call connect() first.",
                details: nil))
            return
        }
        emit(phase: "wifiScanning", deviceId: device.name)
        device.scanWifiList { [weak self] networks, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                if let error = error {
                    result(ErrorMapping.flutterError(forWifiScanError: error))
                    return
                }
                let payload = (networks ?? []).map(self.encode(network:))
                result(payload)
            }
        }
    }

    func provisionWifi(ssid: String, passphrase: String, result: @escaping FlutterResult) {
        guard let device = connectedDevice else {
            result(FlutterError(
                code: "session_failed",
                message: "No active session. Call connect() first.",
                details: nil))
            return
        }
        emit(phase: "applyingCredentials", deviceId: device.name)
        device.provision(ssid: ssid, passPhrase: passphrase) { [weak self] status in
            DispatchQueue.main.async {
                guard let self = self else { return }
                switch status {
                case .success:
                    let r: [String: Any?] = ["status": "success"]
                    self.emit(phase: "finished", deviceId: device.name, result: r)
                    result(r)
                case .configApplied:
                    // Interim phase — ESPProvision will follow up with
                    // .success or .failure. Treat as a progress event;
                    // do NOT resolve the Dart future here.
                    self.emit(phase: "applyingCredentials",
                              deviceId: device.name,
                              message: "configApplied")
                case .failure(let error):
                    if case .wifiStatusDisconnected = error {
                        // Transport-level breakdown: surfaces as an
                        // exception, not an in-band result, per the
                        // sealed-exception contract.
                        result(FlutterError(
                            code: "wifi_provisioning_failed",
                            message: error.description,
                            details: nil))
                        self.emit(phase: "disconnected",
                                  deviceId: device.name,
                                  message: error.description)
                        return
                    }
                    let r = ErrorMapping.provisioningResultMap(for: error)
                    self.emit(phase: "finished", deviceId: device.name, result: r)
                    result(r)
                }
            }
        }
    }

    func sendCustomData(endpoint: String, data: FlutterStandardTypedData, result: @escaping FlutterResult) {
        guard let device = connectedDevice else {
            result(FlutterError(
                code: "session_failed",
                message: "No active session. Call connect() first.",
                details: nil))
            return
        }
        device.sendData(path: endpoint, data: data.data) { response, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(ErrorMapping.flutterError(
                        forSessionError: error,
                        security: self.activeSecurity,
                        ssidOrName: device.name))
                    return
                }
                result(FlutterStandardTypedData(bytes: response ?? Data()))
            }
        }
    }

    func disconnect(result: @escaping FlutterResult) {
        if let device = connectedDevice {
            device.disconnect()
            emit(phase: "disconnected", deviceId: device.name)
        }
        connectedDevice = nil
        result(nil)
    }

    // MARK: - Internals

    private func startBleScan(
        devicePrefix: String,
        timeoutMs: Int,
        result: @escaping FlutterResult
    ) {
        scanInFlight = true
        scanResult = result
        emit(phase: "scanStarted")
        ESPProvisionManager.shared.searchESPDevices(
            devicePrefix: devicePrefix,
            transport: .ble,
            security: .secure2
        ) { [weak self] devices, error in
            DispatchQueue.main.async {
                self?.handleScanCompletion(devices: devices, error: error)
            }
        }
        // ESPProvisionManager's internal scan timer is hardcoded to 5s. We
        // install our own as a fallback in case the SDK's callback never
        // fires (observed once on iOS 17 simulator BLE stack stalls).
        installScanTimer(timeoutMs: timeoutMs)
    }

    private func installScanTimer(timeoutMs: Int) {
        scanTimer?.invalidate()
        let interval = max(0.5, Double(timeoutMs) / 1000.0)
        scanTimer = Timer.scheduledTimer(withTimeInterval: interval,
                                         repeats: false) { [weak self] _ in
            guard let self = self else { return }
            DispatchQueue.main.async {
                guard self.scanInFlight else { return }
                ESPProvisionManager.shared.stopESPDevicesSearch()
                self.emit(phase: "scanFinished", message: "timeout")
                self.scanResult?(FlutterError(
                    code: "device_not_found",
                    message: "BLE scan timed out after \(timeoutMs) ms",
                    details: ["deviceId": ""]))
                self.scanResult = nil
                self.scanInFlight = false
            }
        }
    }

    private func stopScanTimer() {
        scanTimer?.invalidate()
        scanTimer = nil
    }

    private func handleScanCompletion(devices: [ESPDevice]?, error: ESPDeviceCSSError?) {
        guard scanInFlight else { return }
        stopScanTimer()
        scanInFlight = false
        emit(phase: "scanFinished")

        if let error = error {
            switch error {
            case .espDeviceNotFound:
                // No devices found in scan window — return empty list
                // rather than throwing, since this is a routine UI state.
                discoveredDevices = [:]
                scanResult?([])
            default:
                scanResult?(FlutterError(
                    code: "session_failed",
                    message: error.description,
                    details: nil))
            }
            scanResult = nil
            return
        }

        let list = devices ?? []
        discoveredDevices = Dictionary(uniqueKeysWithValues:
            list.map { (key: $0.name, value: $0) })
        scanResult?(list.map(encode(device:)))
        scanResult = nil
    }

    private func encode(device: ESPDevice) -> [String: Any?] {
        return [
            "id": device.name,
            "name": device.name,
            "transport": "ble",
            "rssi": nil,
            "serviceUuid": nil,
            "bssid": nil,
        ]
    }

    private func encode(network: ESPWifiNetwork) -> [String: Any?] {
        let bssidString: String? = network.bssid.isEmpty
            ? nil
            : network.bssid.map { String(format: "%02x", $0) }.joined(separator: ":")
        let authValue: Int
        switch network.auth {
        case .open: authValue = 0
        case .wep: authValue = 1
        case .wpaPsk: authValue = 2
        case .wpa2Psk: authValue = 3
        case .wpaWpa2Psk: authValue = 4
        case .wpa2Enterprise: authValue = 5
        case .wpa3Psk: authValue = 6
        case .wpa2Wpa3Psk: authValue = 7
        case .UNRECOGNIZED(let v): authValue = v
        }
        return [
            "ssid": network.ssid,
            "rssi": Int(network.rssi),
            "authMode": authValue,
            "channel": Int(network.channel),
            "bssid": bssidString,
        ]
    }

    private func emit(
        phase: String,
        deviceId: String? = nil,
        message: String? = nil,
        result: [String: Any?]? = nil
    ) {
        var payload: [String: Any?] = ["phase": phase]
        if let deviceId = deviceId { payload["deviceId"] = deviceId }
        if let message = message { payload["message"] = message }
        if let result = result { payload["result"] = result }
        DispatchQueue.main.async {
            self.eventEmitter?.emit(payload)
        }
    }
}

// MARK: - PoP delegate

/// Supplies ESPProvision with the caller's Proof-of-Possession + the
/// security2 username. A new instance is created per `connect()` so that
/// the delegate captures the exact credentials supplied for that session
/// (rather than relying on mutable bridge state).
private final class PopDelegate: NSObject, ESPDeviceConnectionDelegate {
    let pop: String
    let username: String

    init(pop: String, username: String) {
        self.pop = pop
        self.username = username
    }

    func getProofOfPossesion(forDevice: ESPDevice,
                             completionHandler: @escaping (String) -> Void) {
        completionHandler(pop)
    }

    func getUsername(forDevice: ESPDevice,
                     completionHandler: @escaping (String?) -> Void) {
        completionHandler(username)
    }
}

File: android/src/main/kotlin/com/rainybit/esp_provisioning_flutter/ProvisioningBridge.kt
package com.rainybit.esp_provisioning_flutter

import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.bluetooth.le.ScanResult
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.espressif.provisioning.DeviceConnectionEvent
import com.espressif.provisioning.ESPConstants
import com.espressif.provisioning.ESPDevice
import com.espressif.provisioning.ESPProvisionManager
import com.espressif.provisioning.WiFiAccessPoint
import com.espressif.provisioning.listeners.BleScanListener
import com.espressif.provisioning.listeners.ProvisionListener
import com.espressif.provisioning.listeners.ResponseListener
import com.espressif.provisioning.listeners.WiFiScanListener
import io.flutter.plugin.common.MethodChannel
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * Receives lifecycle events from the bridge. Implemented by the plugin
 * class so the bridge stays decoupled from FlutterEventChannel itself
 * (and so a fake can be substituted in unit tests).
 */
internal interface EspEventEmitter {
    fun emit(event: Map<String, Any?>)
}

/**
 * Default username for ESP-IDF security2 firmware. Production fleets may
 * override this in firmware; if that happens we'd plumb a `username`
 * parameter through `connect()` on the Dart side. Today every device we
 * ship uses the stock value.
 */
private const val DEFAULT_SECURITY2_USERNAME = "wifiprov"

/**
 * Wraps Espressif's `ESPProvisionManager` / `ESPDevice` callback API
 * behind a single per-plugin state machine: discovered-device cache,
 * in-flight scan + connect guards, EventBus subscription for connection
 * lifecycle, and main-thread marshalling of every callback.
 *
 * Mirrors the iOS-side `ProvisioningBridge.swift` 1:1 in semantics. Not
 * thread-safe on its own — every public method must be invoked from the
 * Flutter UI thread (which the method channel guarantees).
 */
internal class ProvisioningBridge(
    private val applicationContext: Context,
    private val eventEmitter: EspEventEmitter
) {

    private val mainHandler = Handler(Looper.getMainLooper())
    private val provisionManager: ESPProvisionManager =
        ESPProvisionManager.getInstance(applicationContext)

    /** Discovered devices from the last scan, keyed by MAC address. The
     *  BluetoothDevice plus the primary service UUID together let us
     *  reconstruct an ESPDevice for connect() without redoing the scan. */
    private val discoveredDevices = mutableMapOf<String, DiscoveredEntry>()

    /** Holds the active ESPDevice between connect → disconnect. Released
     *  on disconnect so EventBus references and the SDK's internal
     *  retained singleton don't keep us alive. */
    private var connectedDevice: ESPDevice? = null

    /** Security used for the current/last connection, needed to make
     *  security-aware decisions in [ErrorMapping] (sec1/sec2 connect
     *  failure → pop_invalid; sec0 → session_failed). */
    private var activeSecurity: ESPConstants.SecurityType = ESPConstants.SecurityType.SECURITY_2

    // Concurrency guards.
    private var scanInFlight = false
    private var connectInFlight = false
    private var scanResult: MethodChannel.Result? = null
    private var connectResult: MethodChannel.Result? = null
    private var pendingConnectDeviceId: String = ""
    private var scanTimeoutRunnable: Runnable? = null
    private var eventBusRegistered = false

    init {
        ensureEventBusRegistered()
    }

    // MARK: - Lifecycle

    /** Tear-down hook invoked by the plugin's onDetachedFromEngine. Drops
     *  EventBus subscription, cancels any in-flight scan, and releases
     *  the connected device. */
    fun dispose() {
        cancelScanTimer()
        if (scanInFlight) {
            try {
                provisionManager.stopBleScan()
            } catch (_: Throwable) { /* best effort */ }
            scanInFlight = false
            scanResult = null
        }
        if (eventBusRegistered) {
            try {
                EventBus.getDefault().unregister(this)
            } catch (_: Throwable) { /* best effort */ }
            eventBusRegistered = false
        }
        connectedDevice = null
        discoveredDevices.clear()
    }

    private fun ensureEventBusRegistered() {
        if (!eventBusRegistered) {
            EventBus.getDefault().register(this)
            eventBusRegistered = true
        }
    }

    // MARK: - Method channel entry points

    @SuppressLint("MissingPermission") // Probe runs first; SDK enforces.
    fun scanBleDevices(devicePrefix: String, timeoutMs: Int, result: MethodChannel.Result) {
        // Cancel any previous in-flight scan first — supersede pattern.
        if (scanInFlight) {
            try {
                provisionManager.stopBleScan()
            } catch (_: Throwable) { /* best effort */ }
            cancelScanTimer()
            // Previous caller gets an empty list rather than an exception;
            // superseding scans is a normal UI pattern.
            scanResult?.success(emptyList<Map<String, Any?>>())
            scanResult = null
            scanInFlight = false
            emit(phase = "scanFinished", message = "superseded by new scan")
        }

        when (val avail = BluetoothStateProbe.check(applicationContext)) {
            BluetoothAvailability.PoweredOff -> {
                result.error(
                    "ble_unavailable",
                    "Bluetooth is off or unsupported on this device",
                    null
                )
                return
            }
            is BluetoothAvailability.PermissionDenied -> {
                result.error(
                    "permission_denied",
                    "Runtime BLE permission '${avail.permission}' is not granted",
                    mapOf("permission" to avail.permission)
                )
                return
            }
            BluetoothAvailability.Available -> Unit
        }

        scanInFlight = true
        scanResult = result
        discoveredDevices.clear()
        emit(phase = "scanStarted")

        provisionManager.searchBleEspDevices(devicePrefix, object : BleScanListener {
            override fun scanStartFailed() {
                mainHandler.post {
                    if (!scanInFlight) return@post
                    cancelScanTimer()
                    scanInFlight = false
                    emit(phase = "scanFinished", message = "scanStartFailed")
                    scanResult?.error(
                        "session_failed",
                        "Failed to start BLE scan (adapter rejected the request)",
                        null
                    )
                    scanResult = null
                }
            }

            override fun onPeripheralFound(device: BluetoothDevice, scanResult: ScanResult) {
                mainHandler.post {
                    val address = try {
                        device.address
                    } catch (_: SecurityException) {
                        return@post // missing BLUETOOTH_CONNECT — should have been caught earlier
                    }
                    val advName: String? = scanResult.scanRecord?.deviceName
                        ?: try { device.name } catch (_: SecurityException) { null }
                    val primaryUuid = scanResult.scanRecord?.serviceUuids?.firstOrNull()?.uuid?.toString()
                    val rssi = scanResult.rssi
                    discoveredDevices[address] = DiscoveredEntry(
                        bluetoothDevice = device,
                        primaryServiceUuid = primaryUuid,
                        advertisedName = advName ?: address,
                        rssi = rssi
                    )
                }
            }

            override fun scanCompleted() {
                mainHandler.post {
                    if (!scanInFlight) return@post
                    cancelScanTimer()
                    scanInFlight = false
                    emit(phase = "scanFinished")
                    scanResult?.success(discoveredDevices.values.map { it.toFlutterMap() })
                    scanResult = null
                }
            }

            override fun onFailure(e: Exception) {
                mainHandler.post {
                    if (!scanInFlight) return@post
                    cancelScanTimer()
                    scanInFlight = false
                    emit(phase = "scanFinished", message = e.message)
                    scanResult?.error(
                        "session_failed",
                        e.message ?: "BLE scan failed",
                        null
                    )
                    scanResult = null
                }
            }
        })

        // SDK's internal scan timer is hardcoded; our additional timer is a
        // defense-in-depth fallback for environments where the listener's
        // scanCompleted callback never fires (observed on certain OEM
        // Android skins).
        installScanTimer(timeoutMs)
    }

    fun stopBleScan(result: MethodChannel.Result) {
        if (scanInFlight) {
            try {
                provisionManager.stopBleScan()
            } catch (_: Throwable) { /* best effort */ }
            cancelScanTimer()
            scanInFlight = false
            emit(phase = "scanFinished", message = "stopped by caller")
            scanResult?.success(discoveredDevices.values.map { it.toFlutterMap() })
            scanResult = null
        }
        result.success(null)
    }

    @SuppressLint("MissingPermission")
    fun connect(
        deviceMap: Map<String, Any?>,
        proofOfPossession: String,
        security: Int,
        result: MethodChannel.Result
    ) {
        if (connectInFlight || connectedDevice != null) {
            result.error(
                "session_failed",
                "A connection is already in flight or established; call disconnect() first",
                null
            )
            return
        }

        val deviceId = deviceMap["id"] as? String
        if (deviceId.isNullOrEmpty()) {
            result.error(
                "session_failed",
                "connect: device map missing valid 'id'",
                null
            )
            return
        }

        val entry = discoveredDevices[deviceId]
        if (entry == null) {
            result.error(
                "device_not_found",
                "Device '$deviceId' is no longer in the scan cache. Re-scan and try again.",
                mapOf("deviceId" to deviceId)
            )
            return
        }

        val espSecurity = securityFromInt(security)
        activeSecurity = espSecurity
        connectInFlight = true
        connectResult = result
        pendingConnectDeviceId = deviceId
        ensureEventBusRegistered()

        val device = provisionManager.createESPDevice(
            ESPConstants.TransportType.TRANSPORT_BLE,
            espSecurity
        )
        device.bluetoothDevice = entry.bluetoothDevice
        device.deviceName = entry.advertisedName
        device.primaryServiceUuid = entry.primaryServiceUuid
        device.proofOfPossession = proofOfPossession
        if (espSecurity == ESPConstants.SecurityType.SECURITY_2) {
            device.userName = DEFAULT_SECURITY2_USERNAME
        }

        emit(phase = "connecting", deviceId = deviceId)
        emit(phase = "sessionEstablishing", deviceId = deviceId)

        device.connectToDevice()
        // Resolution arrives asynchronously via onDeviceConnectionEvent.
    }

    fun scanWifiNetworks(result: MethodChannel.Result) {
        val device = connectedDevice
        if (device == null) {
            result.error(
                "session_failed",
                "No active session. Call connect() first.",
                null
            )
            return
        }
        emit(phase = "wifiScanning", deviceId = device.deviceName)
        device.scanNetworks(object : WiFiScanListener {
            override fun onWifiListReceived(wifiList: ArrayList<WiFiAccessPoint>) {
                mainHandler.post {
                    result.success(wifiList.map { encodeNetwork(it) })
                }
            }

            override fun onWiFiScanFailed(e: Exception) {
                mainHandler.post {
                    val err = ErrorMapping.wifiScanError(e)
                    result.error(err.code, err.message, err.details)
                }
            }
        })
    }

    fun provisionWifi(ssid: String, passphrase: String, result: MethodChannel.Result) {
        val device = connectedDevice
        if (device == null) {
            result.error(
                "session_failed",
                "No active session. Call connect() first.",
                null
            )
            return
        }
        emit(phase = "applyingCredentials", deviceId = device.deviceName)
        device.provision(ssid, passphrase, object : ProvisionListener {
            override fun createSessionFailed(e: Exception) {
                mainHandler.post {
                    val err = ErrorMapping.connectError(e, activeSecurity, device.deviceName ?: "")
                    result.error(err.code, err.message, err.details)
                }
            }

            override fun wifiConfigSent() {
                emit(phase = "applyingCredentials",
                    deviceId = device.deviceName,
                    message = "wifiConfigSent")
            }

            override fun wifiConfigFailed(e: Exception) {
                mainHandler.post {
                    val err = ErrorMapping.provisioningChannelError(e)
                    result.error(err.code, err.message, err.details)
                }
            }

            override fun wifiConfigApplied() {
                emit(phase = "applyingCredentials",
                    deviceId = device.deviceName,
                    message = "wifiConfigApplied")
            }

            override fun wifiConfigApplyFailed(e: Exception) {
                mainHandler.post {
                    val err = ErrorMapping.provisioningChannelError(e)
                    result.error(err.code, err.message, err.details)
                }
            }

            override fun provisioningFailedFromDevice(
                failureReason: ESPConstants.ProvisionFailureReason
            ) {
                mainHandler.post {
                    val resultMap = ErrorMapping.provisioningResultMap(failureReason)
                    emit(phase = "finished",
                        deviceId = device.deviceName,
                        resultPayload = resultMap)
                    result.success(resultMap)
                }
            }

            override fun deviceProvisioningSuccess() {
                mainHandler.post {
                    val resultMap: Map<String, Any?> = mapOf("status" to "success")
                    emit(phase = "finished",
                        deviceId = device.deviceName,
                        resultPayload = resultMap)
                    result.success(resultMap)
                }
            }

            override fun onProvisioningFailed(e: Exception) {
                mainHandler.post {
                    val err = ErrorMapping.provisioningChannelError(e)
                    result.error(err.code, err.message, err.details)
                }
            }
        })
    }

    fun sendCustomData(endpoint: String, data: ByteArray, result: MethodChannel.Result) {
        val device = connectedDevice
        if (device == null) {
            result.error(
                "session_failed",
                "No active session. Call connect() first.",
                null
            )
            return
        }
        device.sendDataToCustomEndPoint(endpoint, data, object : ResponseListener {
            override fun onSuccess(returnData: ByteArray) {
                mainHandler.post { result.success(returnData) }
            }

            override fun onFailure(e: Exception) {
                mainHandler.post {
                    val err = ErrorMapping.customDataError(e)
                    result.error(err.code, err.message, err.details)
                }
            }
        })
    }

    fun disconnect(result: MethodChannel.Result) {
        val device = connectedDevice
        if (device != null) {
            try {
                device.disconnectDevice()
            } catch (_: Throwable) { /* best effort */ }
            emit(phase = "disconnected", deviceId = device.deviceName)
        }
        connectedDevice = null
        result.success(null)
    }

    // MARK: - EventBus subscriber

    /** Receives connect / disconnect lifecycle events posted by the SDK
     *  for the active ESPDevice. ThreadMode.MAIN ensures we are already
     *  on the Flutter UI thread when this fires. */
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onDeviceConnectionEvent(event: DeviceConnectionEvent) {
        when (event.eventType) {
            ESPConstants.EVENT_DEVICE_CONNECTED -> {
                if (connectInFlight) {
                    val device = provisionManager.espDevice
                    connectedDevice = device
                    val deviceId = pendingConnectDeviceId
                    emit(phase = "sessionEstablished", deviceId = deviceId)
                    connectInFlight = false
                    connectResult?.success(null)
                    connectResult = null
                    pendingConnectDeviceId = ""
                }
            }
            ESPConstants.EVENT_DEVICE_CONNECTION_FAILED -> {
                if (connectInFlight) {
                    val deviceId = pendingConnectDeviceId
                    emit(phase = "disconnected", deviceId = deviceId)
                    val err = ErrorMapping.connectFailedWithoutCause(activeSecurity)
                    connectInFlight = false
                    connectResult?.error(err.code, err.message, err.details)
                    connectResult = null
                    pendingConnectDeviceId = ""
                    connectedDevice = null
                }
            }
            ESPConstants.EVENT_DEVICE_DISCONNECTED -> {
                if (connectInFlight) {
                    // Disconnect arrived mid-handshake → reject connect.
                    val deviceId = pendingConnectDeviceId
                    emit(phase = "disconnected", deviceId = deviceId)
                    val err = ErrorMapping.unsolicitedDisconnect(
                        "Device disconnected before session established")
                    connectInFlight = false
                    connectResult?.error(err.code, err.message, err.details)
                    connectResult = null
                    pendingConnectDeviceId = ""
                    connectedDevice = null
                } else if (connectedDevice != null) {
                    // Unsolicited disconnect after we were already connected.
                    emit(phase = "disconnected", deviceId = connectedDevice?.deviceName)
                    connectedDevice = null
                }
            }
        }
    }

    // MARK: - Helpers

    private fun installScanTimer(timeoutMs: Int) {
        cancelScanTimer()
        val runnable = Runnable {
            if (!scanInFlight) return@Runnable
            try {
                provisionManager.stopBleScan()
            } catch (_: Throwable) { /* best effort */ }
            scanInFlight = false
            emit(phase = "scanFinished", message = "timeout")
            scanResult?.success(discoveredDevices.values.map { it.toFlutterMap() })
            scanResult = null
        }
        scanTimeoutRunnable = runnable
        mainHandler.postDelayed(runnable, timeoutMs.toLong().coerceAtLeast(500))
    }

    private fun cancelScanTimer() {
        scanTimeoutRunnable?.let { mainHandler.removeCallbacks(it) }
        scanTimeoutRunnable = null
    }

    private fun encodeNetwork(ap: WiFiAccessPoint): Map<String, Any?> {
        return mapOf(
            "ssid" to ap.wifiName,
            "rssi" to ap.rssi,
            "authMode" to ap.security,
            "channel" to null,
            "bssid" to null,
        )
    }

    private fun securityFromInt(v: Int): ESPConstants.SecurityType {
        return when (v) {
            0 -> ESPConstants.SecurityType.SECURITY_0
            1 -> ESPConstants.SecurityType.SECURITY_1
            else -> ESPConstants.SecurityType.SECURITY_2
        }
    }

    private fun emit(
        phase: String,
        deviceId: String? = null,
        message: String? = null,
        resultPayload: Map<String, Any?>? = null
    ) {
        val payload = mutableMapOf<String, Any?>("phase" to phase)
        if (deviceId != null) payload["deviceId"] = deviceId
        if (message != null) payload["message"] = message
        if (resultPayload != null) payload["result"] = resultPayload
        if (Looper.myLooper() == Looper.getMainLooper()) {
            eventEmitter.emit(payload)
        } else {
            mainHandler.post { eventEmitter.emit(payload) }
        }
    }
}

/** Snapshot of the data we capture when a device is found in a BLE scan,
 *  retained until either the next scan or `dispose()`. */
private data class DiscoveredEntry(
    val bluetoothDevice: BluetoothDevice,
    val primaryServiceUuid: String?,
    val advertisedName: String,
    val rssi: Int,
) {
    @SuppressLint("MissingPermission")
    fun toFlutterMap(): Map<String, Any?> {
        return mapOf(
            "id" to bluetoothDevice.address,
            "name" to advertisedName,
            "transport" to "ble",
            "rssi" to rssi,
            "serviceUuid" to primaryServiceUuid,
            "bssid" to null,
        )
    }
}
